### PR TITLE
[feat] refines the auth flows

### PR DESCRIPTION
**Pull request by pragyakallanagoudar**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

### :palm_tree: What's new in this PR
[//]: # "REQUIRED - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

This PR:
- adds a confirm password input for the signup flow
- provides feedback to the user when they attempt to sign up with an email that is already registered
- adds "new password is not the same as old password" requirement to the reset password PasswordComplexity

### :evergreen_tree: Screenshots
[//]: # "REQUIRED for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy! If you are making changes here, please CC: @kyrenetam at the bottom."

(1) password in confirm password input does not match the original 
passwords do not match

(2) email for the account already exists
email already exists

(3) new password is not the same as old password (working on screenshot 🔨)

## :deciduous_tree: How to review
[//]: # 'REQUIRED - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?'

1. Try to sign up for a new account with an email that already exists, and make sure it errors.
2. Try to sign up for a new account with an email that isn't attached to an existing account. Make the confirm password different from the password input, and make sure it errors.
3. Try to reset your password for an account and enter in the existing password again as the new password. Make sure it errors.

## :seedling: Next steps 
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."

1. Did not add the feature that checks if an account with the email exists for the forgot password page. This is non-trivial because we can't access all the emails in the auth table, so will need to do more digging on this!
2. For the Log In flow, both the email and password text inputs have error messages that are currently not supported ("Email not found" and "Password is incorrect"). It may be useful to add implement a fix for these messages to be displayed in a future sprint. Note: While these messages are not displayed given this has not been implemented, there is a general error message ("Something went wrong. Please try again.") displayed near the top of the card. 

### :information_source: Online sources
[//]: # 'Optional - copy links to any tutorials or documentation that was useful to you when working on this PR'

https://github.com/orgs/supabase/discussions/1282#discussioncomment-5230475

### :potted_plant: Related PRs
[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

[//]: # 'This tags the project leader as a default. Feel free to change, or add on anyone who you should be in on the conversation.'
CC: @varortz

### CODE REVIEW

File: src/app/(auth)/login/page.tsx
Comment (jinkang-0) [inline src/app/(auth)/login/page.tsx:1]: if you have time while working on auth flows, could you wrap each page with forms (i.e. login, sign up) with an actual `<form>` component, and add `type="submit"` to the continue button?

it would have to involve preventing default form submits in the on submit so it doesn't hard reload the page
maybe something like this:
```typescript
const onSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
  e.preventDefault();
  
  /* rest of the code */
}
```

this just allows users to press enter to submit (instead of manually clicking the button)

File: src/api/supabase/queries/password.ts
Comment (varortz) [inline src/api/supabase/queries/password.ts:3]: Can we update the documentation for this function?

Comment (pragyakallanagoudar) [reply src/api/supabase/queries/password.ts:3]: omg im so silly....yes

File: src/app/(auth)/login/page.tsx
Comment (jinkang-0) [reply src/app/(auth)/login/page.tsx:1]: example:
```tsx
export default function Example() {

  const onSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    /* rest of the code */
  }

  return (
    <form onSubmit={onSubmit}>
      <button type="submit">submit</button>
    </form>
  );
}
```

File: src/app/(auth)/confirm-reset-password/page.tsx
Comment (varortz) [inline src/app/(auth)/confirm-reset-password/page.tsx]: can we change all `SpacerDiv`s to `Flex` from `styles/container.ts`, and remove `SpacerDiv` from the styles file?

File: src/app/(auth)/forgot-password/page.tsx
Comment (jinkang-0) [inline src/app/(auth)/forgot-password/page.tsx]: note: remember to switch back before merging

File: src/app/(auth)/forgot-password/page.tsx
Comment (jinkang-0) [inline src/app/(auth)/forgot-password/page.tsx]: nit: can we turn this into a ternary statement?

File: src/app/(auth)/reset-password/page.tsx
Comment (jinkang-0) [inline src/app/(auth)/reset-password/page.tsx]: can we turn this into a ternary statement? if `canReset` is not true, then we should probably add a text here to tell the user

File: src/components/PasswordComplexity.tsx
Comment (jinkang-0) [inline src/components/PasswordComplexity.tsx]: are we using `complexity` anywhere? if not, we could probably just remove this

File: src/app/(auth)/login/page.tsx
Comment (pragyakallanagoudar) [reply src/app/(auth)/login/page.tsx:1]: thank you for the comment + example Jinkang!

File: src/app/(auth)/reset-password/page.tsx
Comment (pragyakallanagoudar) [reply src/app/(auth)/reset-password/page.tsx]: @kyrenetam would it be possible to design a screen for this?

Comment (pragyakallanagoudar) [reply src/app/(auth)/reset-password/page.tsx]: for now imma put null there

File: src/app/(auth)/forgot-password/page.tsx
Comment (jinkang-0) [inline src/app/(auth)/forgot-password/page.tsx]: nit: might be cleaner to do `emailSentCount === 0` since it is technically a number...

File: src/app/(auth)/email-verified/page.tsx
Comment (jinkang-0) [inline src/app/(auth)/email-verified/page.tsx]: nit: we can change the URL to `/onboarding` or use the variable `CONFIG.onboardingHome` so it can be easily refactored later

File: src/app/(auth)/forgot-password/page.tsx
Comment (jinkang-0) [inline src/app/(auth)/forgot-password/page.tsx]: For uncaught error, we might want to say "Something went wrong. Please try again later." instead of displaying some cryptic mumbo jumbo error message given by Supabase.

File: src/app/(auth)/forgot-password/page.tsx
Comment (jinkang-0) [inline src/app/(auth)/forgot-password/page.tsx]: i think there's a regular `BlueButton` somewhere... correct me if im wrong
otherwise, make sure you set `$tertiaryColor` to `COLORS.blueDarker` 🤓

File: src/app/(auth)/reset-password/page.tsx
Comment (jinkang-0) [inline src/app/(auth)/reset-password/page.tsx:50]: similar case here: we may want to just say "Something went wrong. Please try again later."

File: src/app/(auth)/signup/page.tsx
Comment (jinkang-0) [inline src/app/(auth)/signup/page.tsx]: it may be cleaner to turn this into a ternary statement (like forgot password)

File: src/app/(auth)/signup/page.tsx
Comment (jinkang-0) [inline src/app/(auth)/signup/page.tsx]: similar: i think there's a `BlueButton` we can replace this with, or otherwise we should set `$tertiaryColor` to `COLORS.blueDarker`

File: src/components/PasswordComplexity.tsx
Comment (jinkang-0) [inline src/components/PasswordComplexity.tsx]: should we use destructuring here for consistency?

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: is there a reason why we decided to transition away from `AuthResponse`?

we might not need it for sign up, but it could ultimately benefit the future dev team if this sign up wrapper provides the same API return values as a regular Supabase sign up function.

File: src/app/(auth)/forgot-password/page.tsx
Comment (varortz) [reply src/app/(auth)/forgot-password/page.tsx]: yup let's go ahead use `BlueButton`, should be in `components/Buttons.tsx`!

Comment (varortz) [reply src/app/(auth)/forgot-password/page.tsx]: wait shouldn't we use `BigBlueButton` instead?

File: src/app/(auth)/signup/page.tsx
Comment (varortz) [reply src/app/(auth)/signup/page.tsx]: let's use `BigBlueButton`!

File: src/app/(auth)/forgot-password/page.tsx
Comment (jinkang-0) [reply src/app/(auth)/forgot-password/page.tsx]: oh ye, that prolly match the design better

File: src/utils/AuthProvider.tsx
Comment (varortz) [reply src/utils/AuthProvider.tsx]: @pragyakallanagoudar correct me if I'm wrong, but I it seems like the type was updated to fit the solution starting on line 108 taken from https://github.com/orgs/supabase/discussions/1282#discussioncomment-5230475. Could we update the code so that we go back to using `AuthResponse`? This is the structure of an `AuthResponse`, we should be able to directly update its error message:

image
image

File: src/utils/AuthProvider.tsx
Comment (varortz) [inline src/utils/AuthProvider.tsx]: could we change this back to `const value = ...` so that we can return an `AuthResponse` as we were doing previously?

File: src/utils/AuthProvider.tsx
Comment (varortz) [inline src/utils/AuthProvider.tsx]: instead of creating `authError`, we could do something like the following to directly update `value` (`AuthResponse`):

        const authError = new AuthError(
          'A user account with this email already exists',
        );
        value.error = authError;

File: src/utils/AuthProvider.tsx
Comment (varortz) [inline src/utils/AuthProvider.tsx]: we should be able to remove this after refactoring the above code so that we can return an `AuthResponse`! Let's also make sure to return `value` as we were doing previously.

File: src/utils/AuthProvider.tsx
Comment (varortz) [inline src/utils/AuthProvider.tsx]: We'll also need to update these to `value.data.user....` (related to `AuthResponse` changes)

File: src/app/(auth)/signup/page.tsx
Comment (varortz) [inline src/app/(auth)/signup/page.tsx:53]: Since we want to return an `AuthResponse` again, let's revert back to the previous code:
`const { error } = ...`

File: src/app/(auth)/forgot-password/page.tsx
Comment (pragyakallanagoudar) [reply src/app/(auth)/forgot-password/page.tsx]: bet

File: src/app/(auth)/forgot-password/page.tsx
Comment (pragyakallanagoudar) [reply src/app/(auth)/forgot-password/page.tsx]: done!

File: src/components/PasswordComplexity.tsx
Comment (pragyakallanagoudar) [reply src/components/PasswordComplexity.tsx]: i destructured i am so cool

File: src/app/(auth)/forgot-password/page.tsx
Comment (jinkang-0) [inline src/app/(auth)/forgot-password/page.tsx]: is there a reason why we switched to this? it looks like the design has "Could not find email"
image

File: src/app/(auth)/login/page.tsx
Comment (jinkang-0) [inline src/app/(auth)/login/page.tsx]: nit: i think this is capitalized in the design (i.e. "Log In")

File: src/app/(auth)/signup/page.tsx
Comment (jinkang-0) [inline src/app/(auth)/signup/page.tsx]: nit: from design, this should be regular cased (i.e. "Log in")

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: we can replace this somewhat hacky solution with a call to `check_email_exists` (which now works after changing the security level to Definer), which could make it cleaner and easier for future dev team to refactor if they decide to change or add login methods

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: shall we remove this comment?

File: src/app/(auth)/forgot-password/page.tsx
Comment (varortz) [reply src/app/(auth)/forgot-password/page.tsx]: Good catch! I think "Invalid email" is only used on the Sign Up page. Can we change this back to "Could not find email"

File: src/app/(auth)/login/page.tsx
Comment (varortz) [inline src/app/(auth)/login/page.tsx]: Could we update the error messages to match the design:
image

Comment (varortz) [reply src/app/(auth)/login/page.tsx]: Quick note: forgot password flow has a similar error message for the email text input ("Could not find email"). Just messaged @kyrenetam to see if we wanna decide on one of the messages for consistency, will update the comments once there's a decision.

File: src/components/PasswordComplexity.tsx
Comment (varortz) [inline src/components/PasswordComplexity.tsx]: I believe it should be `greyMid` instead of `grayMid`

File: src/app/(auth)/forgot-password/page.tsx
Comment (pragyakallanagoudar) [reply src/app/(auth)/forgot-password/page.tsx]: fixed!

File: src/app/(auth)/signup/page.tsx
Comment (pragyakallanagoudar) [reply src/app/(auth)/signup/page.tsx]: sometimes you scare me jinkang (thanks)

File: src/app/(auth)/login/page.tsx
Comment (pragyakallanagoudar) [reply src/app/(auth)/login/page.tsx]: sg! i'm gonna hold off on any changes for now then

Comment (varortz) [reply src/app/(auth)/login/page.tsx]: from Kyrene: go with "Email not found"

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: nit: dont think these comments are true anymore, we should be able to get rid of them

File: src/utils/AuthProvider.tsx
Comment (jinkang-0) [inline src/utils/AuthProvider.tsx]: nit: might be cleaner if we store the awaited result in a variable first and use it here

**Review by jinkang-0 — APPROVED** [review]
looks good pragya! i just had some minor nits, but otherwise everything looks great - major improvement to the auth flows!

File: src/app/(auth)/forgot-password/page.tsx
Comment (varortz) [inline src/app/(auth)/forgot-password/page.tsx]: The error messages in the designs have no punctuation. Could we remove the period here?
image

File: src/app/(auth)/login/page.tsx
Comment (varortz) [inline src/app/(auth)/login/page.tsx]: Could we also get rid of the periods for these error messages?

File: src/app/(auth)/signup/page.tsx
Comment (varortz) [inline src/app/(auth)/signup/page.tsx]: same here! let's remove the period to remain consistent with the designs.

File: src/app/(auth)/login/page.tsx
Comment (varortz) [reply src/app/(auth)/login/page.tsx]: OH also, are you still down to implement these error checks like we had mentioned the other day?

**Review by varortz — APPROVED** [review]
LOOKS AMAZING PRAGYA! THANKS FOR ENHANCING THE AUTH FLOWS :))

**Pull request merged**
